src/Features/Blockcore.Features.Wallet/Database/WalletStore.cs

```diff
@@ -267,129 +280,395 @@ public IEnumerable<WalletHistoryData> GetAccountHistory(int accountIndex, bool e
+                "and SpendingDetailsTransactionId is null",
+                new { address });
+
+            return trxs.Select(this.Convert);
         }
 
         public WalletBalanceResult GetBalanceForAddress(string address, bool excludeColdStake)
         {
-            var transactions = this.trxCol
-              .Find(Query.And(Query.EQ("Address", new BsonValue(address)), Query.EQ("SpendingDetails", BsonValue.Null)))
-              .Where(x => excludeColdStake && this.network.Consensus.IsProofOfStake ? (x.IsColdCoinStake != true) : true)
-              .GroupBy(x => x.BlockHeight != null)
-              .Select(o =>
-                new
-                {
-                    Confirmed = o.Key,
-                    Amount = o.Sum(x => x.Amount)
-                }).ToList();
+            string excludeColdStakeSql = excludeColdStake && this.network.Consensus.IsProofOfStake ? "AND (IsColdCoinStake = false OR IsColdCoinStake is null) " : string.Empty;
+
+            var sql = "SELECT " +
+                      "BlockHeight as Confirmed," +
+                      "SUM(Amount) as Total " +
+                      "FROM TransactionData " +
+                      $"WHERE SpendingDetailsTransactionId is null AND Address = '{address}' " +
+                      $"{excludeColdStakeSql}" +
+                      $"GROUP BY BlockHeight is not null";
+
+            var result = this.sqliteConnection.Query(sql);
 
             var walletBalanceResult = new WalletBalanceResult();
 
-            foreach (var transaction in transactions)
+            foreach (dynamic item in result)
             {
-                if (transaction.Confirmed == false)
-                    walletBalanceResult.AmountUnconfirmed = transaction.Amount;
-                else
-                    walletBalanceResult.AmountConfirmed = transaction.Amount;
+                if (item.Confirmed == null) walletBalanceResult.AmountUnconfirmed = (long)item.Total;
+                else walletBalanceResult.AmountConfirmed = (long)item.Total;
             }
+
             return walletBalanceResult;
         }
 
         public WalletBalanceResult GetBalanceForAccount(int accountIndex, bool excludeColdStake)
         {
-            var transactions = this.trxCol
-              .Find(Query.And(Query.EQ("AccountIndex", new BsonValue(accountIndex)), Query.EQ("SpendingDetails", BsonValue.Null)))
-              .Where(x => excludeColdStake && this.network.Consensus.IsProofOfStake ? (x.IsColdCoinStake != true) : true)
-              .GroupBy(x => x.BlockHeight != null)
-              .Select(o =>
-                new
-                {
-                    Confirmed = o.Key,
-                    Amount = o.Sum(x => x.Amount)
-                }).ToList();
+            string excludeColdStakeSql = excludeColdStake && this.network.Consensus.IsProofOfStake ? "AND (IsColdCoinStake = false OR IsColdCoinStake is null) " : string.Empty;
+
+            var sql = "SELECT " +
```

Comment:
no need to concatenate queries this way, an easy way is to use @ prefix on string (and a bit on indentation like this helps too)

```
sql = @$"SELECT
            BlockHeight as Confirmed,
            SUM(Amount) as Total
            FROM TransactionData
        WHERE SpendingDetailsTransactionId is null AND Address = '{address}'
            {excludeColdStakeSql}
        GROUP BY BlockHeight is not null";
```

Reply:
What? how didn't I know this

Reply:
way better, isn't it? :P